File: LeetCode/Problem 150.py
# LeetCode: Problem 150
# Jonathan Kosasih

#Evaluate the value of an arithmetic expression in Reverse Polish Notation.
#
#Valid operators are +, -, *, /. Each operand may be an integer or another expression.
#
#Note:
#Division between two integers should truncate toward zero.
#The given RPN expression is always valid.
#That means the expression would always evaluate to a result and there won't be any divide by zero operation.
#------------------------------------------------------------------
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        if(not tokens):
            return 0
        stack = []
        for token in tokens:
            if(token == "+"):
                tempOne = stack.pop()
                tempTwo = stack.pop()
                stack.append(tempTwo + tempOne)
            elif(token == "-"):
                tempOne = stack.pop()
                tempTwo = stack.pop()
                stack.append(tempTwo - tempOne)
            elif(token == "*"):
                tempOne = stack.pop()
                tempTwo = stack.pop()
                stack.append(tempTwo * tempOne)
            elif(token == "/"):
                tempOne = stack.pop()
                tempTwo = stack.pop()
                stack.append(int(tempTwo / tempOne))
            else:
                stack.append(int(token))
        return stack.pop()
        
sol = Solution()
example = ["2", "1", "+", "3", "*"]
print(str(sol.evalRPN(example)))
example = ["4", "13", "5", "/", "+"]
print(str(sol.evalRPN(example)))
example = ["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]
print(str(sol.evalRPN(example)))
example = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"]
print(str(sol.evalRPN(example)))

File: LeetCode/C++/Problem 150.cpp
// LeetCode: Problem 150
// Jonathan Kosasih
/*
Evaluate the value of an arithmetic expression in Reverse Polish Notation.

Valid operators are +, -, *, /. Each operand may be an integer or another expression.

Note:

Division between two integers should truncate toward zero.
The given RPN expression is always valid. That means the expression would always evaluate to a result and there won't be any divide by zero operation.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class Solution {
public:
	int evalRPN(vector<string>& tokens) {
		if (tokens.empty()) {
			return 0;
		}
		stack<int> st;
		for (string s : tokens) {
			if (s == "+") {
				int tempOne = st.top();
				st.pop();
				int tempTwo = st.top();
				st.pop();
				st.push(tempTwo + tempOne);
			}
			else if (s == "-") {
				int tempOne = st.top();
				st.pop();
				int tempTwo = st.top();
				st.pop();
				st.push(tempTwo - tempOne);
			}
			else if (s == "/") {
				int tempOne = st.top();
				st.pop();
				int tempTwo = st.top();
				st.pop();
				st.push(tempTwo / tempOne);
			}
			else if (s == "*") {
				int tempOne = st.top();
				st.pop();
				int tempTwo = st.top();
				st.pop();
				st.push(tempTwo * tempOne);
			}
			else {
				st.push(std::stoi(s));
			}
		}
		return st.top();
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}